src/bp/core/services/action/local-action-server.ts

```diff
@@ -0,0 +1,144 @@
+import bodyParser from 'body-parser'
+import { Logger } from 'botpress/sdk'
+import { ActionDefinition, ActionParameterDefinition } from 'common/typings'
+import { Config } from 'core/app'
+import { ConfigProvider } from 'core/config/config-loader'
+import { BadRequestError, UnauthorizedError } from 'core/routers/errors'
+import { AUDIENCE } from 'core/routers/sdk/utils'
+import { TYPES } from 'core/types'
+import express, { NextFunction, Request, Response } from 'express'
+import { inject, injectable, tagged } from 'inversify'
+import Joi from 'joi'
+import jsonwebtoken from 'jsonwebtoken'
+import _ from 'lodash'
+
+import { BotService } from '../bot-service'
+
+import ActionService from './action-service'
+import { HTTP_ACTIONS_PARAM_TYPES } from './utils'
+
+const _validateRunRequest = (botService: BotService) => async (req: Request, res: Response, next: NextFunction) => {
+  const { appSecret } = await Config.getBotpressConfig()
+
+  try {
+    await Joi.validate(
+      req.body,
+      Joi.object().keys({
+        incomingEvent: Joi.object().required(),
+        actionArgs: Joi.object().required(),
+        actionName: Joi.string().required(),
+        botId: Joi.string().required(),
+        token: Joi.string().required()
+      })
+    )
+  } catch (err) {
+    return next(new BadRequestError(err))
+  }
+
+  const { botId, token } = req.body
+
+  if (!(await botService.botExists(botId))) {
+    return next(new BadRequestError('Unexisting bot'))
+  }
+
+  try {
+    jsonwebtoken.verify(token, appSecret, { audience: AUDIENCE })
+  } catch (err) {
+    return next(new UnauthorizedError('Invalid token'))
+  }
+
+  next()
+}
+
+const _validateListActionsRequest = (botService: BotService) => async (
+  req: Request,
+  res: Response,
+  next: NextFunction
+) => {
+  try {
+    await Joi.validate(
+      req.params,
+      Joi.object().keys({
+        botId: Joi.string().required()
+      })
+    )
+  } catch (err) {
+    return next(new BadRequestError(err))
+  }
+
+  const { botId } = req.params
+
+  if (!(await botService.botExists(botId))) {
+    return next(new BadRequestError('Unexisting bot'))
+  }
```

Comment:
This introduces 2 problems:
1) `forBot()` needs to be `async`, introducing code like this:
```ts
const scopedActionService = await this.actionService.forBot(botId)
const hasAction = await scopedActionService.hasAction(actionName)
```
2) Now every time `forBot()` is called inside a express handler, I will have to wrap the `forBot()` call with a `try... catch` and return a 400 return code, which could have been done easily inside the validation middleware:
```ts
    this.app.post('/action/run', _validateRunRequest, async (req, res) => {
      const { incomingEvent, actionArgs, actionName, botId, token } = req.body

      let scopedActionService
      try {
        scopedActionService = await this.actionService.forBot(botId)
      } catch (e) {
        if (e === 'Unexisting bot') {
          res.sendStatus(400)
          return
        }
      }

      await scopedActionService.runLocalAction({ actionName, actionArgs, incomingEvent, token, runType: 'http' })

      res.send({ incomingEvent })
    })
```

Since existing bots are cached inside the `BotService`, calling `botService.botExists()` introduces no overhead. I think this validation should be done inside the express validation middleware, and not the main express handler.

Reply:
Fixed in a4546d9